### PR TITLE
refactor: `get_finality_updates_from` in eth light client

### LINKED ISSUE #177: Missing finality update caching in `get_finality_updates_from`
https://github.com/synapseweb3/forcerelay/blob/afc197085d32be5aa1dbb243b3ccac1a992e1417/crates/relayer/src/light_client/eth.rs#L717-L761

It appears that we are returning fetched updates immediately when the number of fetched updates is less than expected, but in this case, we're not caching these updates as we would if we exited the loop.

https://github.com/synapseweb3/forcerelay/blob/afc197085d32be5aa1dbb243b3ccac1a992e1417/crates/relayer/src/light_client/eth.rs#L739-L742

### PR DESCRIPTION
**Pull request by whfuyn**
Closes: #177

## Description

Refactor `get_finality_updates_from` for better understanding and a better fix for #177.

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Forcerelay) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/relayer/src/light_client/eth.rs [hunk 1]
```diff
@@ -716,47 +716,41 @@ impl LightClient {
...
+                    .collect::<Vec<_>>();
+                let fetched_updates = futures::future::try_join_all(futs)
+                    .await
+                    .map_err(|e| Error::rpc_response(e.to_string()))?
+                    .into_iter()
+                    .flatten();
+                updates.extend(fetched_updates);
+
+                let fetched = updates.len() as u64;
+                if fetched == 0 {
```
Comment (whfuyn) [inline crates/relayer/src/light_client/eth.rs]: I think this line should be `fetched < n` to match the semantic of the original implementation.

File: crates/relayer/src/light_client/eth.rs [hunk 2]
```diff
@@ -716,47 +716,41 @@ impl LightClient {
...
+                    .collect::<Vec<_>>();
+                let fetched_updates = futures::future::try_join_all(futs)
+                    .await
+                    .map_err(|e| Error::rpc_response(e.to_string()))?
+                    .into_iter()
+                    .flatten();
+                updates.extend(fetched_updates);
+
+                let fetched = updates.len() as u64;
+                if fetched < n {
```
Comment (whfuyn) [inline crates/relayer/src/light_client/eth.rs:742]: It's wrong. `fetched` is the number of total updates fetched so far.

**Comment by ashuralyk** [comment]
it seems work ok I think

**Review by ashuralyk — APPROVED** [review]

**Review by ImJeremyHe — APPROVED** [review]
cool

**Pull request merged**